crates/wdl-engine/src/cache.rs

```diff
@@ -0,0 +1,635 @@
+    cache_dir: Arc<PathBuf>,
+    /// The file transferer that can be used for calculating remote file
+    /// digests.
+    transferer: Arc<dyn Transferer>,
+}
+
+impl State {
+    /// Gets the path to anb entry in the cache given the [`Key`].
+    fn entry_path(&self, key: &Key) -> PathBuf {
+        self.cache_dir.join(key.as_str())
+    }
+}
+
+/// Represents information about content within a call cache entry.
+#[derive(Debug, Default, Serialize, Deserialize)]
+struct Content {
+    /// The location of the content.
+    ///
+    /// May be a local path or a remote URL.
+    location: String,
+    /// The digest of the content, as a hex string.
+    digest: ArrayString<64>,
+}
+
+impl Content {
+    /// Constructs a new [`Content`] from the given evaluation path.
+    ///
+    /// The content digest of the path will be calculated.
+    async fn from_evaluation_path(
+        transferer: &dyn Transferer,
+        path: EvaluationPath,
+        kind: ContentKind,
+    ) -> Result<Self> {
+        let digest = path.calculate_digest(transferer, kind).await?;
+        Ok(Self {
+            location: path.try_into()?,
+            digest: digest.to_hex(),
+        })
+    }
+
+    /// Converts the [`Content`] to an evaluation path.
+    ///
+    /// Returns an error if the current digest of the location does not match
+    /// the stored digest.
+    async fn to_evaluation_path(
+        &self,
+        transferer: &dyn Transferer,
+        kind: ContentKind,
+    ) -> Result<EvaluationPath> {
+        let path: EvaluationPath = self.location.parse()?;
+        let digest = path.calculate_digest(transferer, kind).await?;
+        if digest.to_hex() != self.digest {
+            bail!(
+                "cached content `{location}` was modified",
+                location = self.location
+            );
+        }
+
+        Ok(path)
+    }
+}
+
+/// Represents the serialization of a call cache entry.
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub struct CallCacheEntry {
+    /// The monotonic version of the cache entry.
+    version: u32,
+    /// The digest of the command's evaluated task.
+    command: ArrayString<64>,
+    /// The container used by the task.
+    container: String,
+    /// The shell used by the task.
+    shell: String,
+    /// The requirement digests of the task.
+    requirements: HashMap<String, ArrayString<64>>,
+    /// The hint digests of the task.
+    hints: HashMap<String, ArrayString<64>>,
+    /// The input digests of the task.
+    inputs: HashMap<String, ArrayString<64>>,
+    /// The task's last exit code.
+    exit: i32,
+    /// The task's last stdout content.
+    stdout: Content,
+    /// The task's last stderr content.
+    stderr: Content,
+    /// The task's last work directory content.
+    work: Content,
+}
+
+/// Represents a key for a [`CallCache`].
+///
+/// Additionally stores digests used to validate cache entries.
```

Comment:
There should probably be more of an explanation of what this means somewhere in the comments or at least a link to the "Call Cache Hit" section of the RFC.